### ISSUE TITLE
不支持多标签分类任务蒸馏的原因

### ISSUE THREAD

**Issue by codemayq** [airaria/TextBrewer#50]
非常感谢前辈的工作，这是一个非常标准的蒸馏框架，最近也在持续学习和使用。

不过看文档中写了，不支持 multiLabel ，想问一下不支持的原因是什么，感觉技术理论上应该没有限制。

还是说可能需要单独写一个 distilller 来对特殊的 loss 和label 做适配

**Comment by airaria** [comment]
现有的kd_ce_loss针对多标签分类的输出计算的损失是错误的。

假设多标签分类模型最后输出的tensor形状是 (batch_size, num_classes)，样本属于每个class的概率可以通过在最后一维上做sigmoid得到。

可以用如下的多标签蒸馏loss替换kd_ce_loss:
```python
def kd_multilabel_loss(logits_S, logits_T, temperature=1):
    '''
    Calculate the multi-label binary cross-entropy between logits_S and logits_T.
    We assume that sigmoid(logits_T[i,k]) gives the probability of the i-th example being in the k-th class.

    :param logits_S: Tensor of shape (batch_size, length, num_classes) or (batch_size, num_classes)
    :param logits_T: Tensor of shape (batch_size, length, num_classes) or (batch_size, num_classes)
    '''
    beta_logits_T = logits_T / temperature
    beta_logits_S = logits_S / temperature

    p_T = torch.sigmoid(beta_logits_T)
    log_p_S1 = F.logsigmoid(beta_logits_S)
    log_p_S0 = F.logsigmoid(-beta_logits_S)

    loss = -(p_T * log_p_S1 + (1-p_T) * log_p_S0).mean()

    return loss
```
把上面这段代码放在textbrewer/losses.py，同时修改textbrewer/presets.py中的KD_LOSS_MAP
```python
KD_LOSS_MAP={'mse': kd_mse_loss, 'ce': kd_ce_loss, 'multi-label': kd_multilabel_loss}
```

在初始化DistillationConfig时指定kd_loss_type='multi-label'应该就可以多标签蒸馏了。

**Comment by codemayq** [comment]
> 现有的kd_ce_loss针对多标签分类的输出计算的损失是错误的。
> 
> 假设多标签分类模型最后输出的tensor形状是 (batch_size, num_classes)，样本属于每个class的概率可以通过在最后一维上做sigmoid得到。
> 
> 可以用如下的多标签蒸馏loss替换kd_ce_loss:
> 
> ```python
> def kd_multilabel_loss(logits_S, logits_T, temperature=1):
>     '''
>     Calculate the multi-label binary cross-entropy between logits_S and logits_T.
>     We assume that sigmoid(logits_T[i,k]) gives the probability of the i-th example being in the k-th class.
> 
>     :param logits_S: Tensor of shape (batch_size, length, num_classes) or (batch_size, num_classes)
>     :param logits_T: Tensor of shape (batch_size, length, num_classes) or (batch_size, num_classes)
>     '''
>     beta_logits_T = logits_T / temperature
>     beta_logits_S = logits_S / temperature
> 
>     p_T = torch.sigmoid(beta_logits_T)
>     log_p_S1 = F.logsigmoid(beta_logits_S)
>     log_p_S0 = F.logsigmoid(-beta_logits_S)
> 
>     loss = -(p_T * log_p_S1 + (1-p_T) * log_p_S0).mean()
> 
>     return loss
> ```
> 
> 把上面这段代码放在textbrewer/losses.py，同时修改textbrewer/presets.py中的KD_LOSS_MAP
> 
> ```python
> KD_LOSS_MAP={'mse': kd_mse_loss, 'ce': kd_ce_loss, 'multi-label': kd_multilabel_loss}
> ```
> 
> 在初始化DistillationConfig时指定kd_loss_type='multi-label'应该就可以多标签蒸馏了。

非常感谢，确实多标签的loss 处理会不太一样
